e2e_test/src/rabbitmq_e2e.rs

```diff
@@ -0,0 +1,176 @@
+#![allow(unused_imports)]
+#![allow(unused_variables)]
+use futures::{StreamExt, TryStreamExt};
+use k8s_openapi::api::apps::v1::StatefulSet;
+use k8s_openapi::api::core::v1::{ConfigMap, Pod};
+use k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition;
+use kube::Resource;
+use kube::{
+    api::{
+        Api, AttachParams, AttachedProcess, DeleteParams, ListParams, Patch, PatchParams,
+        PostParams, ResourceExt,
+    },
+    core::crd::CustomResourceExt,
+    Client, CustomResource,
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::thread;
+use std::time::{Duration, Instant};
+use tokio::time::sleep;
+
+use crate::error::Error;
+
+#[derive(kube::CustomResource, Debug, Clone, Deserialize, Serialize, schemars::JsonSchema)]
+#[kube(group = "anvil.dev", version = "v1", kind = "RabbitmqCluster")]
+#[kube(shortname = "rbmq", namespaced)]
+pub struct RabbitmqClusterSpec {
+    pub replicas: i32,
+    #[serde(rename = "rabbitmq", skip_serializing_if = "Option::is_none")]
+    pub rabbitmq_config: Option<RabbitmqClusterConfigurationSpec>,
+}
+#[derive(Debug, Clone, Deserialize, Serialize, schemars::JsonSchema)]
+pub struct RabbitmqClusterConfigurationSpec {
+    // #[serde(rename = "additionalPlugins", skip_serializing_if = "Option::is_none")]
+    // pub additional_plugins: Option<Vec<String>>,
+    #[serde(rename = "additionalConfig", skip_serializing_if = "Option::is_none")]
+    pub additional_config: Option<String>,
+    #[serde(rename = "advancedConfig", skip_serializing_if = "Option::is_none")]
+    pub advanced_config: Option<String>,
+    #[serde(rename = "envConfig", skip_serializing_if = "Option::is_none")]
+    pub env_config: Option<String>,
+}
+
+pub async fn rabbitmq_e2e_test() -> Result<(), Error> {
+    // check if the CRD is already registered
+    let client = Client::try_default().await?;
+    let crds: Api<CustomResourceDefinition> = Api::all(client.clone());
+    let rabbitmq_crd = crds.get("rabbitmqclusters.anvil.dev").await;
+    match rabbitmq_crd {
+        Err(e) => {
+            println!("No CRD found, create one before run the e2e test!\n");
+            return Err(Error::CRDGetFailed(e));
+        }
+        Ok(crd) => {
+            println!("CRD found, continue to run the e2e test!\n");
+        }
+    }
+
+    // create a rabbitmq cluster
+    let pp = PostParams::default();
+    let rabbitmq_api: Api<RabbitmqCluster> = Api::namespaced(client.clone(), "default");
+    let rabbitmq_name = "rabbitmq-test";
+    let rabbitmq_sts_name = format!("{}-server", &rabbitmq_name);
+    let rabbitmq_cm_name = format!("{}-server-conf", &rabbitmq_name);
+    let rabbitmq = RabbitmqCluster::new(
```

Comment:
@ZichengMa I would say we should just submit the rabbitmqcluster cr yaml instead of creating the object here. For two reasons: (1) We kind of repeat the definition of the rabbitmqcluster that we already have in deps_hack. So we should just make the e2e test simple and agnostic of the cr definition. (2) the other testing tools that can generate different inputs usually generate a cr yaml, so it is better to align with their output for later use.